pycaqtimage/pycaqtimage.sip

```diff
@@ -0,0 +1,727 @@
+        free(imageBuffer->imageData);
+    free(imageBuffer);
+}  
+
+/*
+ * Parameters:
+ *    imageDisp - The oriented, false-colored QImage to return.
+ *    px_, py_ - numpy arrays (float64) to hold the (oriented) projections.
+ *    image_ - numpy array (uint32) to hold the current (oriented) image.
+ *    orientation - The orientation constant.
+ *
+ * The new regime is that we *immediately* reorient whatever we are
+ * reading in, and all calculations are on the transformed image.
+ */
+PyObject* pyCreateImageBuffer(QImage* imageDisp, PyObject *px_, PyObject *py_, PyObject *image_, int srcwidth, int srcheight, int orientation)
+{
+    PyArrayObject *px = NULL;
+    PyArrayObject *py = NULL;
+    PyArrayObject *image = NULL;
+    int lenx, leny;
+    static int first = 1;
+    if (first) {
+        first = 0;
+        _import_array(); /* SIP chokes on import_array, so directly call the internal routine. */
+    }
+    if (imageDisp == NULL)
+        Py_RETURN_NONE;
+    if (px_ != NULL && PyArray_Check(px_))
+        px = (PyArrayObject *)px_;
+    if (py_ != NULL && PyArray_Check(py_))
+        py = (PyArrayObject *)py_;
+    if (image_ != NULL && PyArray_Check(image_))
+        image = (PyArrayObject *)image_;
+  
+    ImageBuffer* imageBuffer = (ImageBuffer*) malloc(sizeof(ImageBuffer));
+    imageBuffer->srcwidth    = srcwidth;
+    imageBuffer->srcheight   = srcheight;
+    if (orientation & 2) {
+        imageBuffer->imgwidth  = lenx = srcheight;
+        imageBuffer->imgheight = leny = srcwidth;
+    } else {
+        imageBuffer->imgwidth  = lenx = srcwidth;
+        imageBuffer->imgheight = leny = srcheight;
+    }
+    imageBuffer->size        = lenx * leny;
+    imageBuffer->imageDisp   = imageDisp;
+    if (px != NULL && PyArray_NDIM(px) == 1 && 
+        PyArray_DIM(px, 0) == lenx && PyArray_TYPE(px) == NPY_DOUBLE) {
+#ifdef DEBUG
+        printf("Using numpy array for projX\n");
+#endif
+        imageBuffer->projSumX    = (double *)PyArray_DATA(px);
+        imageBuffer->freeX       = 0;
+    } else {
+#ifdef DEBUG
+        printf("Allocating memory for projX\n");
+        printf("DIM(0) = %ld, expected %d, type = %d, expected %d\n",
+	       PyArray_DIM(px, 0), lenx, PyArray_TYPE(px), NPY_DOUBLE);
+#endif
+        imageBuffer->projSumX    = (double *) malloc( lenx * sizeof(double) );
+        imageBuffer->freeX       = 1;
+    }
+    if (py != NULL && PyArray_NDIM(py) == 1 && 
+        PyArray_DIM(py, 0) == leny && PyArray_TYPE(py) == NPY_DOUBLE) {
+#ifdef DEBUG
+        printf("Using numpy array for projY\n");
+#endif
+        imageBuffer->projSumY    = (double *)PyArray_DATA(py);
+        imageBuffer->freeY       = 0;
+    } else {
+#ifdef DEBUG
+        printf("Allocating memory for projY\n");
+        printf("DIM(0) = %ld, expected %d, type = %d, expected %d\n",
+	       PyArray_DIM(py, 0), leny, PyArray_TYPE(py), NPY_DOUBLE);
+#endif
+        imageBuffer->projSumY    = (double *) malloc( leny * sizeof(double) );
```

Comment:
The comments above the function indicate it should be taking in numpy arrays always for `px_` and `py_`; what's  the usage of the other scenario where they aren't specified as such?

Reply:
More debugging than anything else.  This enabled me to still make progress when I wasn't properly allocating the arrays.

I should probably just die if they aren't numpy arrays of the proper type and dimension.

Reply:
I see... I think "numpy array or die" is a reasonable approach here - it would cut down on the complexity at least a bit

Reply:
I can agree.  Fixed.